cmd/encryption-v1.go

```diff
@@ -253,7 +253,9 @@ func DecryptRequest(client io.Writer, r *http.Request, metadata map[string]strin
 		Key: keyEncryptionKey,
 	})
 	if n != 32 || err != nil {
-		return nil, errObjectTampered
+		// Either the provided key does not match or the object was tampered.
+		// To provide strict AWS S3 compatibility we return: access denied.
+		return nil, errSSEKeyMismatch
```

Comment:
should we return `errSSEKeyMismatch`  for `n != 32` and `err != nil` as errObjectTampered?  is it only possible to decrypt the object to know if the key mismatched?

Reply:
We don't decrypt the object here. We decrypt the sealed `object-encryption-key`. The sealed `object-encryption-key` is part of the object metadata and is 64 bytes long. (16b header, 32b key, 16b tag). The decryption of the sealed `object-encryption-key` fails because:
 - The client provided key is not correct: `keyEncryptionKey' = SHA256(key' || iv)`
 - The object metadata was modified (IV): `keyEncryptionKey' = SHA256(key || iv')`
 - The object metadata was modified (sealedKey): `authentication tag mismatch`

`n != 32` is always true if `err != nil` - except the `object-encryption-key` was > 32 during encryption but that would be a bug in the `EncryptRequest` function and should never happen!

Reply:
okay